### DISCUSSION TITLE
Does Vercel support on-premises deployment?

### DISCUSSION THREAD

**Discussion by xiaosa66** [vercel/community#3297] [Feature Question] [answered]
### Question

Does Vercel support on-premises deployment? Paid option is acceptable.

**Comment by swarnava** [comment]
Hi there,

Currently, we don't support Self hosted deployment. You can use Vercel CLI to setup your own CI/CD pipeline to deploy apps: https://vercel.com/guides/how-can-i-use-gitlab-pipelines-with-vercel

**Reply by xiaosa66** [reply]
Thank you very much!

**Reply by xiaosa66** [reply]
It looks like it still requires uploading the code to Vercel, right? Due to security reasons, we cannot provide the source code, but we still want to use Vercel. Is there any solution?

**Reply by leerob** [reply, marked as answer]
If you are using the Vercel CLI with `vercel build` and `vercel deploy --prebuilt`, you are only sending the compiled code to Vercel to run. Vercel would not have access to the source code.

https://vercel.com/docs/cli/build
